google/cloud/storage/signed_url_options.h

```diff
@@ -56,6 +56,104 @@ inline AddExtensionHeaderOption AddExtensionHeader(std::string header,
       std::make_pair(std::move(header), std::move(value)));
 }
 
+/**
+ * Add a extension header to a signed URL.
+ */
+struct AddQueryParameterOption
+    : public internal::ComplexOption<AddQueryParameterOption, std::string> {
+  using ComplexOption<AddQueryParameterOption, std::string>::ComplexOption;
+  static char const* name() { return "query-parameter"; }
+
+  /**
+   * Escapes a string using URL encoding.
+   *
+   * This function acts as a compilation barrier because including
+   * internal/curl_wrappers.h in this header would expose that header (and the
+   * headers it depends on) in the public interface, and we want to avoid that.
+   */
+  static std::string UrlEscape(std::string const& value);
+};
+
+inline AddQueryParameterOption WithAcl() {
+  return AddQueryParameterOption("acl");
+}
+
+inline AddQueryParameterOption WithBilling() {
+  return AddQueryParameterOption("billing");
+}
+
+inline AddQueryParameterOption WithCompose() {
+  return AddQueryParameterOption("compose");
+}
+
+inline AddQueryParameterOption WithCors() {
+  return AddQueryParameterOption("cors");
+}
+
+inline AddQueryParameterOption WithEncryption() {
+  return AddQueryParameterOption("encryption");
+}
+
+inline AddQueryParameterOption WithEncryptionConfig() {
+  return AddQueryParameterOption("encryptionConfig");
+}
+
+inline AddQueryParameterOption WithGeneration(std::uint64_t generation) {
+  return AddQueryParameterOption(
+      "generation=" +
+      AddQueryParameterOption::UrlEscape(std::to_string(generation)));
+}
+
+inline AddQueryParameterOption WithGenerationMarker(std::uint64_t generation) {
```

Comment:
The docs indicate that you should only supply query string parameters that designate a subresource (and, I suppose, the others that are explicitly called out like response-content-disposition, response-content-type... and userProject makes sense as well).

Some of these -- mainly the ones used for listing and composing: `With{Compose,Generation,GenerationMarker,Marker}` -- don't seem like they belong here.

Reply:
`WithCompose` sounds perfectly reasonable, you can create a signed URL that allows you to upload an object, why not a signed URL that allows you to compose an object?

`WithGeneration`: I believe this can be used to read an object for a given generation?  That sounds useful with a signed URL.

`WithGenerationMarker` and `WithMarker` these are used to restrict the output of listing buckets, but I have no idea, happy to remove.

Reply:
I don't think I've ever seen a good use case for Compose-via-signed-url, but I suppose that doesn't mean it isn't possible. Fair point :)

And good point on generation as well -- I've filed a docs bug to change the language on [1]. The query parameters aren't limited to subresources; rather, there are a select group of query params that should not be used.

[1] https://cloud.google.com/storage/docs/access-control/signed-urls#about-canonical-extension-headers

Reply:
@houglum [here's](https://github.com/googleapis/google-cloud-go/issues/1199) a developer use-case. There's an open question about supporting the compose operation with signed URLs in Go and across libraries. At the moment not all clients provide the same level of support introduced in this PR. It looks like it's not necessarily an open question then, but a better design to simplify support in a similar way to the C++ client. I have digressed, thanks for the update @coryan.